File: ASL_Angio_Reader/MIP_reader.m
close all
clear
clc

%% PAR File lines:
% 1-11 = Data Description
% 12-47 = General Information
% 48-52 = Pixel Values
% 53-97 = Image Information Definition
% 98-100 = Header von "Image Information"
% 101-end = real Image part


%% ASL File
%%%%%%%%%%%%%%%%%%%%%%% User Input %%%%%%%%%%%%%%%%%%%%%%%%%%%%%
filename   = 'ASL_10_1';   % Name of file to be opened
%%%%%%%%%%%%%%%%%%%%%%% User Input %%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% %%%%Read image parameter from PAR file%%%%
parrawdata = textscan(fopen([filename '.PAR']),'%s', 'delimiter', '\n');
pardata=parrawdata{1};

% %%%%Read relevant image acquisition parameters%%%%

for cntr=12:47 % reads number of slices
    temp = strread(pardata{cntr},'%s', 'delimiter', ':');
    temp2 = temp{1};
    if (temp2(1) == '#')
        continue % Ignore comments
    end
    if (strfind(temp2,'Max. number of slices/locations')>1)
        NSlices = str2double(temp{2}); 
     end
end

for cntr=12:47 % reads number of dynamics
    temp = strread(pardata{cntr},'%s', 'delimiter', ':');
    temp3 = temp{1};
    if (temp3(1) == '#')
        continue % Ignore comments
    end
    if (strfind(temp3,'Number of label types   <0=no ASL>')>1)
        NDynamics = str2double(temp{2}); 
     end
end

for cntr=12:47 % reads number of dynamics
    temp = strread(pardata{cntr},'%s', 'delimiter', ':');
    temp4 = temp{1};
    if (temp4(1) == '#')
        continue % Ignore comments
    end
    if (strfind(temp4,'Max. number of cardiac phases')>1)
        NPhases = str2double(temp{2}); 
     end
end

% %%%%choose line where the slice information starts%%%%

centr   = 101;
Lastline = size(pardata,1)-1;
%temp   = strread(pardata{cntr},'%s');


% %%%%read out the important values%%%%
% %%%%counting begins after the header (line 101)%%%%
for cntr=101:Lastline
    param = strread(pardata{centr},'%f');
    
    RI          = param(12);
    RS          = param(13);
    SS          = param(14);
    
    xRes        = param(10);
    yRes        = param(11);
    
end
Itype       = 4;
clear parrawdata;
clear pardata;

% %%%%Calculations as described in PAR file:%%%%

% DV = PV * RS + RI
% FP = DV / RS * SS

% %%%%ASLdata: read full data%%%%
recrawdata = fopen([filename '.REC'],'r');

% %%%%"Put" full dataset into an x-dimensional Array seperated by the Imagetypes:%%%%%
% %%%%0 = Magnitude, 1 = Real, 2 = Imaginary, 3 = Phase (not sure...)%%%%%

h = waitbar(0,'Separating Image Types');

for x = 0:Itype
    
   if x == 0
        ASLdatamag  = (reshape(fread(recrawdata,xRes*yRes*NDynamics*NSlices*NPhases,'int16'),xRes,yRes,2,NDynamics/2,NSlices,NPhases));
   end
    waitbar(x/Itype)
end
close(h)

clear recrawdata;
% %%%%The ASLdata is the image structure of the files. The resulting image looks like this: %%%%

% %%%%[Number of Rows, Number of Columns,2, Number of the "Dynamics" (Labeltypes), Slice Number, (Cardiac) Phase number]%%%%


% %%%%MAGNITUDE IMAGES%%%%


% %%%%separate even and odd numbered Labels (odd = tag, even = control)%%%%

even = ASLdatamag(:,:,2,:,:,:);

odd = ASLdatamag(:,:,1,:,:,:);

%recalculate the values as described in PAR file

clear ASLdatamag;

PV_odd = odd*RS+RI;
PV_even = even*RS+RI;
FP_odd = PV_odd/(RS*SS);
FP_even = PV_even/(RS*SS);

clear even;
clear odd;
clear PV_odd;
clear PV_even;
% %%%%preallocating for speed%%%%

even_acc=zeros(xRes,yRes,NSlices,NPhases);
odd_acc=zeros(xRes,yRes,NSlices,NPhases);

% %%%%add up all even and odd images%%%%

for x=1:xRes
    for y=1:yRes
        for z=1:NSlices
            for t=1:NDynamics/2
                for u = 1:NPhases
                even_acc(x,y,z,u)=even_acc(x,y,z,u)+FP_even(x,y,t,1,z,u);
                end
            end
        end
    end
end

for x=1:xRes
    for y=1:yRes
        for z=1:NSlices
            for t=1:NDynamics/2
                for u = 1:NPhases
                odd_acc(x,y,z,u)=odd_acc(x,y,z,u)+FP_odd(x,y,t,1,z,u);
                end
            end
        end
    end
end

% %%%%perform the subtraction%%%%
sub=odd_acc - even_acc;

clear even_acc;
clear odd_acc;

%%%%Save images in Nifti Format%%%%

h = waitbar(0,'Magnitude');

for x = 1:NPhases
    
    waitbar(x/NPhases)
    
    if x == 1
        MIP1 = max(sub(:,:,:,1), [], 3);
        
        origin = [4 4 4]; datatype = 64;
        nii = make_nii(MIP1, [], origin, datatype);
        save_nii(nii,'MIP1.nii');
    end
    
    if x == 2
        MIP2 = max(sub(:,:,:,2), [], 3);
        
        origin = [1 1 1]; datatype = 64;
        nii = make_nii(MIP2, [], origin, datatype);
        save_nii(nii,'MIP2.nii');
    end
    
    if x == 3
        MIP3 = max(sub(:,:,:,3), [], 3);
        
        origin = [1 1 1]; datatype = 64;
        nii = make_nii(MIP3, [], origin, datatype);
        save_nii(nii,'MIP3.nii');
    end
    
    if x == 4
        MIP4 = max(sub(:,:,:,4), [], 3);
        
        origin = [1 1 1]; datatype = 64;
        nii = make_nii(MIP4, [], origin, datatype);
        save_nii(nii,'MIP4.nii');
    end
    
    if x == 5
        MIP5 = max(sub(:,:,:,5), [], 3);
        
        origin = [1 1 1]; datatype = 64;
        nii = make_nii(MIP5, [], origin, datatype);
        save_nii(nii,'MIP5.nii');
        
    end
    
    if x == 6
        MIP6 = max(sub(:,:,:,6), [], 3);
        
        origin = [1 1 1]; datatype = 64;
        nii = make_nii(MIP6, [], origin, datatype);
        save_nii(nii,'MIP6.nii');
        
    end
end

close(h)




%         tform = maketform('affine',[-2.5 0 0; 0 1 0; 0 0 0.5]);
%         MIPx = imtransform(MIP1,tform);
%         origin = [1 1 1]; datatype = 64;
%         nii = make_nii(MIPx, [], origin, datatype);
%         save_nii(nii,'MIPx.nii');
        

File: ASL_Angio_Reader/MIP_reader.py
import re
import numpy as np
from tqdm import trange
import make_nii
import gc


# PAR File lines (0-indexed):
# 0-10 = Data Description
# 11-46 = General Information
# 47-51 = Pixel Values
# 52-96 = Image Information Definition
# 97-99 = Header von "Image Information"
# 100-end = real Image part

def MIP_reader(filename):
    # Read image parameter from PAR file
    with open(filename + '.PAR', 'r') as parfile:
        pardata = parfile.readlines()

    # Read relevant image acquisition parameters
    for curr_line in range(11, 47):
        temp = re.split(r':', pardata[curr_line])
        temp2 = temp[0].strip()
        if temp2.startswith('#'):
            continue  # Ignore comments
        if 'Max. number of slices/locations' in temp2:
            NSlices = int(temp[1])
        if 'Number of label types   <0=no ASL>' in temp2:
            NDynamics = int(temp[1])
        if 'Max. number of cardiac phases' in temp2:
            NPhases = int(temp[1])

    # Choose line where the slice information starts
    curr_line = 100
    last_line = len(pardata)

    # read out the important values
    # counting begins after the header (line 100)
    for curr_line in range(100, last_line):
        param = list(map(float, pardata[curr_line].split()))
        RI = param[11]
        RS = param[12]
        SS = param[13]
        xRes = param[9]
        yRes = param[10]

    Itype = 4
    del(pardata)
    gc.collect() 

    # Read full data
    recdata = open(filename + '.REC', 'r')

    # Create a progress bar for separating image types
    for x in trange(Itype + 1):
        if x == 0:
            # Example data loading and reshaping, replace with actual code
            recdata = np.random.randint(0, 100, size=(xRes * yRes * NDynamics * NSlices * NPhases), dtype=np.int16)
            ASLdatamag = np.reshape(recdata, (xRes, yRes, 2, NDynamics // 2, NSlices, NPhases))

    # Clear recdata
    del recdata
    gc.collect() 

    # Separate even and odd numbered Labels
    even = ASLdatamag[:, :, 1, :, :, :]
    odd = ASLdatamag[:, :, 0, :, :, :]

    # Recalculate values as described in PAR file
    PV_odd = odd * RS + RI
    PV_even = even * RS + RI
    FP_odd = PV_odd / (RS * SS)
    FP_even = PV_even / (RS * SS)

    # Add up all even and odd images
    even_acc = np.zeros((xRes, yRes, NSlices, NPhases))
    odd_acc = np.zeros((xRes, yRes, NSlices, NPhases))

    for x in range(xRes):
        for y in range(yRes):
            for z in range(NSlices):
                for t in range(NDynamics//2):
                    for u in range(NPhases):
                        even_acc[x, y, z, u] += FP_even[x, y, t, 1, z, u]
                        odd_acc[x, y, z, u] += FP_odd[x, y, t, 1, z, u]


    sub = odd_acc - even_acc

    #Create progress bar 
    for x in trange(NPhases):
        datatype = np.int64
        origin = [1, 1, 1]
        if x == 1:
            MIP1 = np.max(sub[:,:,:,0], axis=2)
            origin = [4, 4, 4]
            nii = make_nii(MIP1, None, origin, datatype)
            save_nii(nii,'MIP1.nii')
        if x == 2:
            MIP2 = np.max(sub[:,:,:,1], axis=2)
            nii = make_nii(MIP2, None, origin, datatype)
            save_nii(nii,'MIP2.nii')
        if x == 3:
            MIP3 = np.max(sub[:,:,:,2], axis=2)
            nii = make_nii(MIP3, None, origin, datatype)
            save_nii(nii,'MIP3.nii')
        if x == 4:
            MIP4 = np.max(sub[:,:,:,3], axis=2)
            nii = make_nii(MIP4, None, origin, datatype)
            save_nii(nii,'MIP4.nii')
        if x == 5:
            MIP5 = np.max(sub[:,:,:,4], axis=2)
            nii = make_nii(MIP5, None, origin, datatype)
            save_nii(nii,'MIP5.nii')
        if x == 6:
            MIP6 = np.max(sub[:,:,:,5], axis=2)
            nii = make_nii(MIP6, None, origin, datatype)
            save_nii(nii,'MIP6.nii')
            
